### PR TITLE
Point-2-Point communicator.

### PR DESCRIPTION
**Pull request by dr-robertk**
This PR adds a point-2-point communicator for Dune/OPM. This will be used in an upcoming PR for the parallel output. 

@atgeirr: I have tried to include as many white space violations of the coding style as possible ;-).

### CODE REVIEW

**Comment by atgeirr** [comment]
I assume this is essentially cloned from ALUGrid functionality as we discussed earlier? If so, can I also assume it has been well tested? I am not quite sure how to review this, since I am not an MPI expert. I will take a look though.

> I have tried to include as many white space violations of the coding style as possible

Always a rebel...

File: dune/grid/common/p2pcommunicator.hh [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    /** \brief constructor taking memory reserve estimation factor (default is 1.1, i.e. 10% over estimation )
+     */
+    SimpleMessageBuffer( const double factor = 1.1 )
+      : buffer_(), factor_( factor )
+    {
+      resetReadPosition();
+    }
+
+    /** \brief clear the buffer */
+    void clear() { BufferType().swap( buffer_ ); resetReadPosition(); }
```
Comment (atgeirr) [inline dune/grid/common/p2pcommunicator.hh]: Why is the "hard" clear necessary for buffer_? I would think a plain buffer_.clear() would be enough?

**Comment by dr-robertk** [comment]
The implementation of the  MPI API has indeed been well tested. The SimpleMessageBuffer is new, but the test shows that it works.

File: dune/grid/common/p2pcommunicator.hh (hunk 1 above)
Comment (dr-robertk) [inline dune/grid/common/p2pcommunicator.hh]: Probably. Just a matter of taste. Not very important.

File: dune/grid/common/p2pcommunicator.hh [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+      size_t pos  = buffer_.size();
+      const size_t sizeNeeded = pos + tsize ;
+      // reserve with some 10% overestimation
+      if( buffer_.capacity() < sizeNeeded )
+      {
+        reserve( size_t(factor_ * sizeNeeded) ) ;
+      }
+      // resize to size need to store value
+      buffer_.resize( sizeNeeded );
+      for(unsigned int i=0; i<tsize; ++i )
```
Comment (atgeirr) [inline dune/grid/common/p2pcommunicator.hh]: Could replace this by std::copy().

File: dune/grid/common/p2pcommunicator.hh [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+    /** \brief read value from buffer, value must implement the operator= correctly (i.e. no internal pointers etc.) */
+    template <class T>
+    void read( T& value ) const
+    {
+      // read bytes from stream and store in value
+      const size_t tsize = sizeof( T );
+      union { T value; char bytes[ tsize ]; } convert;
+      assert( pos_ + tsize <= buffer_.size() );
+      for( unsigned int i=0; i<tsize; ++i )
```
Comment (atgeirr) [inline dune/grid/common/p2pcommunicator.hh]: As above. Also: why not use size_t for `i` when that is used for `tsize`?

File: dune/grid/common/p2pcommunicator.hh [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+  /** \brief Point-2-Point communicator for exchange messages between processes */
+  template <class MsgBuffer, class Comm = MPIHelper::MPICommunicator >
+  class Point2PointCommunicator : public CollectiveCommunication< Comm >
+  {
+  protected:
+    typedef CollectiveCommunication< Comm > BaseType;
+    typedef Point2PointCommunicator< MsgBuffer, Comm > ThisType;
+
+    // starting message tag
+    static const int messagetag = 234;
```
Comment (atgeirr) [inline dune/grid/common/p2pcommunicator.hh]: Is 234 an MPI code or an arbitrary number?

File: dune/grid/common/p2pcommunicator.hh [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    };
+
+  public :
+    // default constructor
+    Point2PointCommunicator() : BaseType() { removeLinkage(); }
+    // constructor taking CollectiveCommunication
+    Point2PointCommunicator( const BaseType& comm ) : BaseType( comm ) { removeLinkage(); }
+
+    // return new tag number for the exchange messages
+    static int getMessageTag( const unsigned int icrement )
```
Comment (atgeirr) [inline dune/grid/common/p2pcommunicator.hh]: Typo?

**Comment by bska** [comment]
I'll just state once that the kind of type-punning used in the `read` and `write` functions always makes me uneasy.  Type-punning through a union is supported and portable in C99 and C11 but not in C89.  C++ introduces the notion of an **active** member of a `union` and basically states that reading a union through a member other than the one used to store it is implementation defined.

```C++
template <class T>
void write( const T& value )
{
    // union to access bytes in value
    const size_t tsize = sizeof( T );
    union { T value; char bytes[ tsize ]; } convert;
    convert.value = value;
    // ...
    for(unsigned int i=0; i<tsize; ++i )
    {
        buffer_[ pos++ ] = convert.bytes[ i ] ;
    }
}
```

**Comment by atgeirr** [comment]
> reading a union through a member other than the one used to store it is implementation defined.

Would a reinterpret_cast<char*> behave better (or at least more standard-guaranteed) for this purpose?

**Comment by bska** [comment]
>> reading a union through a member other than the one used to store it is implementation defined.
>
> Would a `reinterpret_cast<char*> behave better (or at least more standard-guaranteed) for this purpose?

Yes, at least as long as `T` is a POD.  I think I'd prefer
```C++
reinterpret_cast<unsigned char*>
```
though.

File: dune/grid/common/p2pcommunicator.hh (hunk 4 above)
Comment (dr-robertk) [inline dune/grid/common/p2pcommunicator.hh]: It's and arbitrary number away from standard MPI message tasks which should be between 0 and 32768.

**Comment by atgeirr** [comment]
Googled this and found no clear recommendations. Apparently there is undefined behaviour in all the methods. Makes you wonder how memcpy can be implemented in a standard-conforming way, maybe it cannot? Or I just did not find a good resource.

**Comment by dr-robertk** [comment]
Actually, reinterpret_cast is used in the ALUGrid implementation and I though that was a bit fishy. 
So I'll change that then.

**Comment by bska** [comment]
> Makes you wonder how `memcpy` can be implemented in a standard-conforming way, maybe it cannot?

It cannot.  Function `memcpy()` is a portable interface into a non-portable feature.  There is almost always some assembly implementation that takes care of marshalling the bits once it gets close to the hardware.

**Comment by dr-robertk** [comment]
I addressed the above comments.

File: dune/grid/common/p2pcommunicator.hh [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        buffer_.reserve( size );
+    }
+
+    /** \brief resize buffer to 'size' entries  */
+    void resize( const size_t size )
+    {
+        buffer_.resize( size );
+    }
+
+    /** \brief write value to buffer, value must implement the operator= correctly (i.e. no internal pointers etc.) */
```
Comment (atgeirr) [inline dune/grid/common/p2pcommunicator.hh]: Final point (I promise): this comment (and below) are now misleading. I guess it should say that T must be a POD?

Comment (dr-robertk) [inline dune/grid/common/p2pcommunicator.hh]: That's actually not entirely correct. As long as the operator= is implemented correctly and the types contained in value contain only types that only contain types that at some point are PODs it's ok. But, constructors and destructors are allowed.

Comment (atgeirr) [inline dune/grid/common/p2pcommunicator.hh]: Maybe "standard layout" is the term I am looking for? Anyway, I'll not let it hold back. Merging.

**Comment by bska** [comment]
This does not build out of the box in an environment that targets Dune 2.2.x.  I can, however, restore the build in that environment, at least in a sequential configuration (#171), if I apply the change below to
```
dune/grid/common/p2pcommunicator.hh
```
(i.e., look for Dune MPI headers in `dune/common` rather than `dune/common/parallel` when targeting 2.2.x).  The `p2pcommunicator_test` also runs successfully in this environment.

Is Dune 2.2.x support worth the added maintenance cost or should we simply disable the point-to-point communicator unless we're targeting Dune 2.3.x or later?
```diff
diff --git a/dune/grid/common/p2pcommunicator.hh b/dune/grid/common/p2pcommunicator.hh
index fd95dad..79e13ca 100644
--- a/dune/grid/common/p2pcommunicator.hh
+++ b/dune/grid/common/p2pcommunicator.hh
@@ -24,13 +24,29 @@
 #include <set>
 #include <map>
 
+#include <dune/common/version.hh>
+
+#if DUNE_VERSION_NEWER(DUNE_COMMON, 2, 3)
+
 #include <dune/common/parallel/mpihelper.hh>
 #include <dune/common/parallel/collectivecommunication.hh>
 
 // the following implementation is only available in case MPI is available
 #if HAVE_MPI
 #include <dune/common/parallel/mpicollectivecommunication.hh>
-#endif
+#endif // HAVE_MPI
+
+#else // DUNE_VERSION_NEWER(DUNE_COMMON, 2, 3)
+
+#include <dune/common/mpihelper.hh>
+#include <dune/common/collectivecommunication.hh>
+
+// the following implementation is only available in case MPI is available
+#if HAVE_MPI
+#include <dune/common/mpicollectivecommunication.hh>
+#endif // HAVE_MPI
+
+#endif // DUNE_VERSION_NEWER(DUNE_COMMON, 2, 3)
```

**Pull request merged**